Funcky/Extensions/EnumerableExtensions/AdjacentGroupBy.cs

```diff
@@ -0,0 +1,242 @@
+
+        /// <summary>
+        /// Groups adjacent elements of a source sequence according to a specified key selector function and creates a result value from each group and its key. Key values are compared by using a specified comparer, and the elements of each group are projected by using a specified function.
+        /// </summary>
+        /// <typeparam name="TSource">Type of the elements in <paramref name="source"/> sequence.</typeparam>
+        /// <typeparam name="TKey">The type of the key returned by keySelector.</typeparam>
+        /// <typeparam name="TElement">The type of the elements in each IGrouping{TKey,TElement}.</typeparam>
+        /// <typeparam name="TResult">The type of the result value returned by resultSelector.</typeparam>
+        /// <param name="source">The source sequence.</param>
+        /// <param name="keySelector">A function to extract the key for each element.</param>
+        /// <param name="elementSelector">A function to map each source element to an element in the IGrouping{TKey,TElement}.</param>
+        /// <param name="resultSelector">A function to create a result value from each group.</param>
+        /// <param name="comparer">An IEqualityComparer{T} to compare keys.</param>
+        /// <returns>A collection of elements of type TResult where each element represents a projection over a group and its key.</returns>
+        [Pure]
+        public static IEnumerable<TResult> AdjacentGroupBy<TSource, TKey, TElement, TResult>(
+            this IEnumerable<TSource> source,
+            Func<TSource, TKey> keySelector,
+            Func<TSource, TElement> elementSelector,
+            Func<TKey, IList<TElement>, TResult> resultSelector,
+            IEqualityComparer<TKey> comparer)
+        {
+            using var enumerator = source.GetEnumerator();
+
+            if (!enumerator.MoveNext())
+            {
+                yield break;
+            }
+
+            var group = new List<TElement>() { elementSelector(enumerator.Current) };
+            var key = keySelector(enumerator.Current);
+
+            while (enumerator.MoveNext())
+            {
+                if (comparer.Equals(key, keySelector(enumerator.Current)))
+                {
+                    group.Add(elementSelector(enumerator.Current));
+                }
+                else
+                {
+                    yield return resultSelector(key, group);
+                    group = new List<TElement>() { elementSelector(enumerator.Current) };
+                    key = keySelector(enumerator.Current);
+                }
+            }
+
+            yield return resultSelector(key, group);
+        }
+
+        internal static Grouping<TKey, TElement> CreateGrouping<TKey, TElement>(TKey key, IList<TElement> elements)
+        {
+            return new Grouping<TKey, TElement>(key, elements);
+        }
+
+        public class Grouping<TKey, TElement> : IGrouping<TKey, TElement>, IList<TElement>
```

Comment:
Why is this class public?

Reply:
Because the References implementation says so: https://github.com/dotnet/runtime/blob/d6887d4aeec045aa4d708463abdfb6e071660dc8/src/libraries/System.Linq/src/System/Linq/Grouping.cs#L42

Reply:
Ah, yes I remember seeing that comment. Please copy the comment from dotnet/runtime